src/com/amazon/corretto/crypto/provider/HkdfSecretKeyFactorySpi.java

```diff
@@ -0,0 +1,166 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.corretto.crypto.provider;
+
+import static com.amazon.corretto.crypto.provider.HkdfSpec.HKDF_EXPAND_MODE;
+import static com.amazon.corretto.crypto.provider.HkdfSpec.HKDF_EXTRACT_MODE;
+import static com.amazon.corretto.crypto.provider.HkdfSpec.HKDF_MODE;
+
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactorySpi;
+import javax.crypto.spec.SecretKeySpec;
+
+class HkdfSecretKeyFactorySpi extends SecretKeyFactorySpi {
+  private final int digestCode;
+  private final int digestLength;
+
+  private HkdfSecretKeyFactorySpi(final int digestCode, final int digestLength) {
+    this.digestCode = digestCode;
+    this.digestLength = digestLength;
+  }
+
+  @Override
+  protected SecretKey engineGenerateSecret(final KeySpec keySpec) throws InvalidKeySpecException {
+    if (!(keySpec instanceof HkdfSpec)) {
+      throw new InvalidKeySpecException("KeySpec must be an instance of HkdfSpec");
+    }
+    final HkdfSpec spec = (HkdfSpec) keySpec;
+
+    final byte[] resultBytes;
+    switch (spec.mode) {
+      case HKDF_MODE:
+        checkExpandLength(spec.desiredLength);
+        resultBytes = new byte[spec.desiredLength];
+        hkdf(
+            resultBytes,
+            resultBytes.length,
+            digestCode,
+            spec.secretOrPrk,
+            spec.secretOrPrk.length,
+            spec.salt,
+            spec.salt.length,
+            spec.info,
+            spec.info.length);
+        break;
+      case HKDF_EXTRACT_MODE:
+        // In extract mode, we ignore the value of desiredLength and use
+        // the output size of the digest as desiredLength
+        resultBytes = new byte[digestLength];
+        hkdfExtract(
+            resultBytes,
+            resultBytes.length,
+            digestCode,
+            spec.secretOrPrk,
+            spec.secretOrPrk.length,
+            spec.salt,
+            spec.salt.length);
+        break;
+      case HKDF_EXPAND_MODE:
+        checkExpandLength(spec.desiredLength);
+        resultBytes = new byte[spec.desiredLength];
+        hkdfExpand(
+            resultBytes,
+            resultBytes.length,
+            digestCode,
+            spec.secretOrPrk,
+            spec.secretOrPrk.length,
+            spec.info,
+            spec.info.length);
+        break;
+      default:
+        throw new AssertionError(
+            "This should not be reachable, since the constructor of HkdfSpec ensures the value is"
+                + " one of the above choices.");
+    }
+
+    return new SecretKeySpec(resultBytes, spec.algorithmName);
+  }
+
+  private void checkExpandLength(final long outLen) {
+    final long upperLimit = ((long) digestLength) * 255L;
```

Comment:
where does the constant 255L come from?

Reply:
I commented the code. It's coming from the standard.